### PR TITLE
OTG Test - RT-5.1: Singleton Interface

### PR DESCRIPTION
**Pull request by octpetre**
OTG version of the RT-5.1: Singleton Interface test case.
@liulk or @bstoll please assign.

### CODE REVIEW

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 1]
```diff
@@ -371,11 +450,11 @@ func (tc *testCase) run(t *testing.T) {
 	for _, c := range []struct {
 		ipName     string
 		shouldFrag bool
-		ipHeader   ondatra.Header
+		ipHeader   string
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:453]: Also add the neighbor watcher as a wait function?

Here: `waitFn func(otg *otg.OTG)`

And provide separate functions for waitOTGIPv4 and waitOTGIPv6?

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 2]
```diff
@@ -246,6 +272,25 @@ func (tc *testCase) verifyATE(t *testing.T) {
 	})
 }
 
+// Waits for an ARP entry to be present on ateSrc
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:275]: This should be two separate functions, one for IPv4 and one for IPv6. See the test case definition below. For brevity, they should just take `otg` as the argument.

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 3]
```diff
@@ -281,26 +327,57 @@ func (tc *testCase) testFlow(t *testing.T, packetSize uint16, ipHeader ondatra.H
...
+	tc.top.Flows().Clear().Items()
+	flow := tc.top.Flows().Add().SetName("Flow")
+	flow.Metrics().SetEnable(true)
+	switch ipHeader {
+	case "IPv4":
+		ipType = "IPv4"
+		flow.TxRx().Device().SetTxNames([]string{i1 + ".IPv4"}).SetRxNames([]string{i2 + ".IPv4"})
+		flow.Size().SetFixed(int32(packetSize))
+		e1 := flow.Packet().Add().Ethernet()
+		e1.Src().SetValue(ateSrc.MAC)
```
[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:339]: Done

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 2 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:275]: Done

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 1 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go]: I created separate functions for IPv4 and IPv6 but I don't understand the request for neighbor watcher as a wait function. Please take a look at the code and let me know

[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go]: I have removed ipHeader from the struct. Please let me know if I need to change anything.

[the remark this answers is not in the document]
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go]: switch removed and config and wait functions are now passed in the test run() function. The tests are now referring this functions in the struct

Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go]: switch removed and config and wait functions are now passed in the test run() function. The tests are now referring this functions in the struct

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 4]
```diff
@@ -246,6 +276,64 @@ func (tc *testCase) verifyATE(t *testing.T) {
...
+	flow.TxRx().Device().SetTxNames([]string{ateSrc.Name + ".IPv4"}).SetRxNames([]string{ateDst.Name + ".IPv4"})
+	flow.Size().SetFixed(int32(packetSize))
+	v4 := flow.Packet().Add().Ipv4()
+	v4.Src().SetValue(ateSrc.IPv4)
+	v4.Dst().SetValue(ateDst.IPv4)
+	tc.ate.OTG().PushConfig(t, tc.top)
+	otg.Telemetry().Interface(ateSrc.Name+".Eth").Ipv4NeighborAny().LinkLayerAddress().Watch(
+		t, time.Minute, func(val *otgtelemetry.QualifiedString) bool {
+			return val.IsPresent()
+		}).Await(t)
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:291]: Remove? Already in waitOTGIPv4ARPEntry.

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 5]
```diff
@@ -246,6 +276,64 @@ func (tc *testCase) verifyATE(t *testing.T) {
...
+	flow := tc.top.Flows().Items()[0]
+	flow.TxRx().Device().SetTxNames([]string{ateSrc.Name + ".IPv6"}).SetRxNames([]string{ateDst.Name + ".IPv6"})
+	v6 := flow.Packet().Add().Ipv6()
+	v6.Src().SetValue(ateSrc.IPv6)
+	v6.Dst().SetValue(ateDst.IPv6)
+	tc.ate.OTG().PushConfig(t, tc.top)
+	otg.Telemetry().Interface(ateSrc.Name+".Eth").Ipv6NeighborAny().LinkLayerAddress().Watch(
+		t, time.Minute, func(val *otgtelemetry.QualifiedString) bool {
+			return val.IsPresent()
+		}).Await(t)
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:320]: Remove?

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 6]
```diff
@@ -246,6 +276,64 @@ func (tc *testCase) verifyATE(t *testing.T) {
...
+	flow.TxRx().Device().SetTxNames([]string{ateSrc.Name + ".IPv4"}).SetRxNames([]string{ateDst.Name + ".IPv4"})
+	v4 := flow.Packet().Add().Ipv4()
+	v4.DontFragment().SetValue(1)
+	v4.Src().SetValue(ateSrc.IPv4)
+	v4.Dst().SetValue(ateDst.IPv4)
+	tc.ate.OTG().PushConfig(t, tc.top)
+	otg.Telemetry().Interface(ateSrc.Name+".Eth").Ipv4NeighborAny().LinkLayerAddress().Watch(
+		t, time.Minute, func(val *otgtelemetry.QualifiedString) bool {
+			return val.IsPresent()
+		}).Await(t)
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:306]: Remove?

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go [hunk 7]
```diff
@@ -246,6 +276,64 @@ func (tc *testCase) verifyATE(t *testing.T) {
...
+}
+
+func (tc *testCase) waitOTGIPv4ARPEntry(t *testing.T) {
+	tc.ate.OTG().Telemetry().Interface(ateSrc.Name+".Eth").Ipv4NeighborAny().LinkLayerAddress().Watch(
+		t, time.Minute, func(val *otgtelemetry.QualifiedString) bool {
+			return val.IsPresent()
+		}).Await(t)
+}
+
+func (tc *testCase) waitOTGIPv6ARPEntry(t *testing.T) {
```
Comment (liulk) [inline feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:330]: ND is the more accurate term for IPv6, not ARP. Or change both IPv4 and IPv6 ARP to just Neighbor?

**Review by liulk — CHANGES_REQUESTED** [review]

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 6 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:306]: Sorry...gone

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 5 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:320]: Forgot about this..sorry

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 4 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:291]: Done

File: feature/interface/singleton/otg_tests/singleton_test/singleton_test.go (hunk 7 above)
Comment (octpetre) [reply feature/interface/singleton/otg_tests/singleton_test/singleton_test.go:330]: Replaced with "Neighbor"

**Review by liulk — APPROVED** [review]

**Pull request merged**